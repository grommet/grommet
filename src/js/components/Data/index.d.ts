import * as React from 'react';
import { BoxProps } from '../Box/index';

export interface View {
  name?: string;
  page?: number; // when paging
  properties?:
    | string[]
    | {
        [property: string]:
          | undefined
          | boolean
          | (string | number | boolean)[]
          | { max: number; min: number };
      };
  search?: string;
  sort?: {
    property: string;
    direction: 'asc' | 'desc';
  };
  step?: number; // page size
  columns?: string[];
}

export interface DataProps {
  id?: string;

  data: object[];
  total?: number;
  // when paging
  filteredTotal?: number;

  defaultView?: View;
  view?: string | View;
  onView?: (view: View) => void;

  // whether to render a Toolbar
  toolbar?: boolean | 'search' | 'filters' | 'view';

  properties?:
    | string[]
    | {
        [key: string]: {
          badge?: boolean;
          filter?: boolean;
          // for DataTable column header, DataFilter label, DataTableColumns label
          label?: string | React.ReactNode;
          // DataFilter options
          options?: (
            | string
            | number
            | {
                label: string;
                value: string | number | boolean;
              }
          )[];
          // DataFilter range
          range?: {
            max?: number;
            min?: number;
            step?: number;
          };
          // for internal filtering only, should searching evaluate this property
          search?: boolean;
          sort?: boolean;
        };
      };

  messages?: {
    dataFilters?: {
      clear?: string;
      heading?: string;
      open?: string;
      openSet?: {
        singular?: string;
        plural?: string;
      };
    };
    dataForm?: {
      submit?: string;
    };
    dataSearch?: {
      label?: string;
      open?: string;
    };
    dataSort?: {
      ascending?: string;
      by?: string;
      descending?: string;
      direction?: string;
      open?: string;
    };
    dataSummary?: {
      filtered?: string;
      filteredSingle?: string;
      items?: string;
      itemsSingle?: string;
      selected?: string;
      total?: string;
      totalSingle?: string;
    };
    dataTableColumns?: {
      open?: string;
      order?: string;
      select?: string;
      tip?: string;
    };
    dataTableGroupBy?: {
      clear?: string;
      label?: string;
    };
    dataView?: {
      label?: string;
    };
  };

  views?: View[];

  // Future selection
  // select: string[];
  // onSelect: (select: string[]) => void;

  // Future manage named views
  // onCreateView?: (view: View) =>  void;
  // onDeleteView?: (string) => void;
}

type divProps = Omit<
  React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'onClick'
>;

export interface DataExtendedProps extends BoxProps, DataProps, divProps {}

declare const Data: React.FC<DataExtendedProps>;

export { Data };
