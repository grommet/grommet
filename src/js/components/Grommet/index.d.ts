import * as React from 'react';
import { ThemeType } from '../../themes';
import { BackgroundType } from '../../utils';

export interface GrommetProps {
  background?: BackgroundType;
  containerTarget?: HTMLElement;
  cssVars?: boolean;
  dir?: 'rtl';
  full?: boolean | 'min';
  options?: {
    box?: {
      cssGap?: boolean;
    };
    layer?: {
      singleId?: boolean;
    };
    drop?: {
      checkContainingBlock?: boolean;
    };
  };
  messages?: {
    messages?: {
      button?: {
        busy?: string;
        success?: string;
      };
      calendar?: {
        previousMove?: string;
        previous?: string;
        nextMove?: string;
        next?: string;
      };
      carousel?: {
        previous?: string;
        next?: string;
        jump?: string;
      };
      dateInput?: {
        openCalendar?: string;
        enterCalendar?: string;
        exitCalendar?: string;
      };
      dataChart?: {
        detailTitle?: string;
        detailFocus?: string;
      };
      dataFilters?: {
        clear?: string;
        heading?: string;
        open?: string;
        openSet?: {
          singular?: string;
          plural?: string;
        };
      };
      dataForm?: {
        submit?: string;
      };
      dataSearch?: {
        label?: string;
        open?: string;
      };
      dataSort?: {
        ascending?: string;
        by?: string;
        descending?: string;
        direction?: string;
        open?: string;
      };
      dataSummary?: {
        filtered?: string;
        filteredSingle?: string;
        items?: string;
        itemsSingle?: string;
        selected?: string;
        total?: string;
        totalSingle?: string;
      };
      dataTable?: {
        ascending?: string;
        collapse?: string;
        collapseAll?: string;
        decrease?: string;
        descending?: string;
        increase?: string;
        expand?: string;
        expandAll?: string;
        resizerAria?: string;
        rows?: string;
        rowsChanged?: string;
        rowsSingle?: string;
        searchBy?: string;
        total?: string;
        totalSingle?: string;
      };
      dataTableColumns?: {
        open?: string;
        order?: string;
        pinned?: string;
        select?: string;
        tip?: string;
        orderAria?: string;
        selectAria?: string;
      };
      dataTableGroupBy?: {
        clear?: string;
        label?: string;
      };
      dataView?: {
        label?: string;
      };
      fileInput?: {
        alert?: {
          maxFile?: string;
          maxSize?: string;
        };
        browse?: string;
        dropPrompt?: string;
        dropPromptMultiple?: string;
        files?: string;
        maxFile?: string;
        maxSizeSingle?: string;
        maxSizeMultiple?: {
          singular?: string;
          plural?: string;
        };
        remove?: string;
        removeAll?: string;
      };
      form?: {
        invalid?: string;
        required?: string;
      };
      formField?: {
        maxCharacters?: {
          remaining?: {
            singular?: string;
            plural?: string;
          };
          overLimit?: {
            singular?: string;
            plural: string;
          };
        };
      };
      list?: {
        pinned?: string;
      };
      menu?: {
        openMenu?: string;
        closeMenu?: string;
      };
      meter?: {
        bar?: {
          singular?: string;
          plural?: string;
        };
        circle?: {
          singular?: string;
          plural?: string;
        };
        pie?: {
          singular?: string;
          plural?: string;
        };
        semicircle?: {
          singular?: string;
          plural?: string;
        };
      };
      notification?: {
        close?: string;
      };
      pagination?: {
        stepLabel?: string;
        summary?: string;
        summaryNoItems?: string;
      };
      rangeSelector?: {
        lower?: string;
        upper?: string;
      };
      tag?: {
        removeLabel?: {
          nameAndValue?: string;
          valueOnly?: string;
        };
      };
      select?: {
        multiple?: string;
        selected?: string;
      };
      selectMultiple?: {
        clearAll?: string;
        clearAllA11y?: string;
        open?: string;
        optionNotSelected?: string;
        optionSelected?: string;
        search?: string;
        selectAll?: string;
        selectAllA11y?: string;
        selectDrop?: string;
        selected?: string;
        selectedOfTotal?: string;
        selectedOptions?: string;
        showMore?: string;
        summarizedValue?: string;
      };
      skipLinks?: {
        skipTo?: string;
      };
      tabs?: {
        tabContents?: string;
      };
      textInput?: {
        enterSelect?: string;
        suggestionsCount?: string;
        suggestionsExist?: string;
        suggestionIsOpen?: string;
      };
      video?: {
        audioDescriptions?: string;
        captions?: string;
        closeMenu?: string;
        description?: string;
        fullScreen?: string;
        progressMeter?: string;
        scrubber?: string;
        openMenu?: string;
        pauseButton?: string;
        playButton?: string;
        volumeDown?: string;
        volumeUp?: string;
      };
    };
    format?: (...args: any[]) => void;
  };
  onAnalytics?: (data: any) => void;
  plain?: boolean;
  theme?: ThemeType;
  themeMode?: 'dark' | 'light' | 'auto';
  userAgent?: string;
}

export interface GrommetExtendedProps
  extends GrommetProps,
    Omit<
      React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLDivElement>,
        HTMLDivElement
      >,
      'dir'
    > {}

declare const Grommet: React.FC<GrommetExtendedProps>;

export { Grommet };
