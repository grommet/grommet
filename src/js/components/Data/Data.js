import React, { useContext, useEffect, useMemo, useState } from 'react';
import { AnnounceContext } from '../../contexts';
import { Box } from '../Box';
import { DataFilters } from '../DataFilters';
import { DataSearch } from '../DataSearch';
import { DataSummary } from '../DataSummary';
import { DataView } from '../DataView';
import { Toolbar } from '../Toolbar';
import { DataContext } from '../../contexts/DataContext';
import { DataPropTypes } from './propTypes';
import { MessageContext } from '../../contexts/MessageContext';
import { filter } from './filter';

const defaultDefaultView = { search: '' };

const normalizeView = (viewProp, views) =>
  (typeof viewProp === 'string' && views?.find((v) => v.name === viewProp)) ||
  (typeof viewProp === 'object' && viewProp);

export const Data = ({
  children,
  data: dataProp = [],
  defaultView = defaultDefaultView,
  filteredTotal,
  id = 'data',
  messages,
  onView,
  properties,
  toolbar,
  total,
  view: viewProp,
  views,
  ...rest
}) => {
  const [view, setView] = useState(
    normalizeView(viewProp || defaultView, views),
  );
  useEffect(
    () => setView(normalizeView(viewProp || defaultView, views)),
    [defaultView, viewProp, views],
  );
  const [toolbarKeys, setToolbarKeys] = useState([]);

  const result = useMemo(() => {
    if (onView)
      // caller is filtering
      return {
        data: dataProp,
        total,
        filteredTotal: filteredTotal ?? dataProp?.length ?? 0,
      };
    return filter(dataProp, view, properties);
  }, [dataProp, filteredTotal, onView, properties, total, view]);

  // used by DataFilters to determine if badge should appear on Filter button
  const [filtersCleared, setFiltersCleared] = useState(true);

  const [selected, setSelected] = useState([]);

  const announce = useContext(AnnounceContext);
  const { format } = useContext(MessageContext);
  // Announce to screen readers when search or filters are
  // applied and affect the underlying result set
  useEffect(() => {
    let messageId;
    if (result.total !== result.filteredTotal) {
      if (result.filteredTotal === 1) messageId = 'dataSummary.filteredSingle';
      else messageId = 'dataSummary.filtered';
    } else if (result.total === 1) messageId = 'dataSummary.totalSingle';
    else messageId = 'dataSummary.total';

    // helps account for cases like 0 results of 1 item
    const items = format({
      id: result.total === 1 ? 'dataSummary.itemsSingle' : 'dataSummary.items',
      messages: messages?.dataSummary,
    });

    announce(
      `${format({
        id: messageId,
        messages: messages?.dataSummary,
        values: {
          filteredTotal: result.filteredTotal,
          total: result.total,
          items,
        },
      })}${
        selected > 0
          ? `, ${format({
              id: 'dataSummary.selected',
              messages: messages?.dataSummary,
              values: {
                selected,
              },
            })}`
          : ''
      }`,
    );
  }, [
    announce,
    format,
    messages?.dataSummary,
    result.filteredTotal,
    result.total,
    selected,
  ]);

  // what we use for DataContext value
  const contextValue = useMemo(() => {
    const value = {
      id,
      messages,
      properties,
      filtersCleared,
      setFiltersCleared,
      selected,
      setSelected,
      view,
      views,
      ...result,
    };

    value.clearFilters = () => {
      const nextView = { ...view };
      delete nextView.properties;
      delete nextView.page;
      // by clearing the properties from a view, it is no
      // longer reflecting the view
      delete nextView.name;
      delete nextView.view;
      setFiltersCleared(true);
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.onView = (nextView) => {
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.addToolbarKey = (key) => {
      setToolbarKeys((prevKeys) => {
        if (prevKeys.includes(key)) return prevKeys;
        return [...prevKeys, key];
      });
    };
    value.toolbarKeys = toolbarKeys;

    return value;
  }, [
    id,
    messages,
    filtersCleared,
    onView,
    properties,
    result,
    selected,
    toolbarKeys,
    view,
    views,
  ]);

  let toolbarContent;
  if (toolbar) {
    toolbarContent = [
      // TO DO theme object
      <Toolbar key="toolbar" gap="medium">
        <Toolbar>
          {(toolbar === true || toolbar === 'search') && <DataSearch />}
          {(toolbar === true || toolbar === 'filters') && <DataFilters layer />}
        </Toolbar>
        {(toolbar === true || toolbar === 'view') && <DataView />}
      </Toolbar>,
      <DataSummary key="summary" />,
    ];
  }

  return (
    <DataContext.Provider value={contextValue}>
      <Box id={id} flex={false} {...rest}>
        {toolbarContent}
        {children}
      </Box>
    </DataContext.Provider>
  );
};

Data.propTypes = DataPropTypes;
