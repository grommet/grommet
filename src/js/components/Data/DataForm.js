import React, {
  useMemo,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { Button } from '../Button';
import { Footer } from '../Footer';
import { Form } from '../Form';
import { DataContext } from '../../contexts/DataContext';
import { DataFormContext } from '../../contexts/DataFormContext';
import { MessageContext } from '../../contexts/MessageContext';
import { useDebounce } from '../../utils/use-debounce';

const MaxForm = styled(Form)`
  max-width: 100%;
  ${(props) => props.fill && 'max-height: 100%;'}
`;

// We convert the view structure to something more flat to work better
// with the Form inputs. These keys are how we flatten the Form value object
// from the view object.
export const formSearchKey = '_search';
export const formSortKey = '_sort';
export const formRangeKey = '_range';
export const formStepKey = '_step';
export const formPageKey = '_page';
export const formColumnsKey = '_columns';
export const formGroupByKey = '_groupBy';
export const formViewNameKey = '_view';

const viewFormKeyMap = {
  search: formSearchKey,
  sort: formSortKey,
  step: formStepKey,
  page: formPageKey,
  columns: formColumnsKey,
  groupBy: formGroupByKey,
  view: formViewNameKey,
};

// flatten nested objects.
// For example: { a: { b: v, c: z } } -> { 'a.b': v, 'a.c': z }
const flatten = (formValue, options) => {
  const result = JSON.parse(JSON.stringify(formValue));
  Object.keys(result).forEach((i) => {
    // We check the type of the i using
    // typeof() function and recursively
    // call the function again
    // ignore _range situations
    if (
      typeof result[i] === 'object' &&
      !Array.isArray(result[i]) &&
      (options?.full || !result[i][formRangeKey])
    ) {
      const temp = flatten(result[i], options);
      Object.keys(temp).forEach((j) => {
        // Store temp in result
        // ignore empty arrays
        if (
          !Array.isArray(temp[j]) ||
          (Array.isArray(temp[j]) && temp[j].length)
        )
          result[`${i}.${j}`] = temp[j];
      });
      delete result[i];
    }
  });
  return result;
};

// unflatten nested objects. For example: { 'a.b': v } -> { a: { b: v } }
const unflatten = (formValue) => {
  const result = JSON.parse(JSON.stringify(formValue));
  const specialKeys = Object.values(viewFormKeyMap);
  Object.keys(result)
    .filter((k) => !specialKeys.includes(k))
    .forEach((k) => {
      const parts = k.split('.');
      const val = result[k];
      delete result[k];
      let parent = result;
      while (parts.length > 1) {
        const sub = parts.shift();
        if (!parent[sub]) parent[sub] = {};
        parent = parent[sub];
      }
      parent[parts.shift()] = val;
    });
  return result;
};

// converts from the external view format to the internal Form value format
const viewToFormValue = (view) => {
  const result = { ...(view?.properties || {}) };
  // convert { min: , max: } range to [min, max] for RangeSelector
  Object.keys(result).forEach((key) => {
    if (
      typeof result[key]?.min === 'number' ||
      typeof result[key]?.max === 'number'
    ) {
      result[key] = { [formRangeKey]: [result[key].min, result[key].max] };
    }
  });

  // convert formal view keys to their form '_' prefixed counterparts
  Object.keys(viewFormKeyMap).forEach((key) => {
    if (view?.[key]) result[viewFormKeyMap[key]] = view[key];
  });
  // always have some blank search text
  if (!result[formSearchKey]) result[formSearchKey] = '';

  if (view?.sort) result[formSortKey] = view.sort;
  if (view?.name) result[formViewNameKey] = view.name;
  if (view?.columns) result[formColumnsKey] = view.columns;
  if (view?.groupBy) result[formGroupByKey] = view.groupBy;

  return unflatten(result);
};

// converts from the internal Form value format to the external view format
const formValueToView = (value, views) => {
  let result = {};

  // if the user chose a view, use that
  if (value[formViewNameKey])
    result = JSON.parse(
      JSON.stringify(views.find((v) => v.name === value[formViewNameKey])),
    );

  const valueCopy = { ...value };

  Object.keys(viewFormKeyMap).forEach((key) => {
    if (valueCopy[viewFormKeyMap[key]]) {
      result[key] = valueCopy[viewFormKeyMap[key]];
    }
    delete valueCopy[viewFormKeyMap[key]];
  });

  // flatten any nested objects
  const flatValue = flatten(valueCopy);

  result.properties = { ...(result.properties || {}), ...flatValue };

  // convert any ranges
  Object.keys(result.properties).forEach((key) => {
    if (result.properties[key][formRangeKey]) {
      result.properties[key] = {
        min: result.properties[key][formRangeKey][0],
        max: result.properties[key][formRangeKey][1],
      };
    }
  });

  return result;
};

// remove any empty arrays of property values by deleting the key for
// that property in the view properties
const clearEmpty = (formValue, pendingReset) => {
  const value = formValue;
  Object.keys(value).forEach((k) => {
    if (Array.isArray(value[k]) && value[k].length === 0) delete value[k];
    // special case for when range selector returns to its min/max
    // flat format with full: true needed to match filter name structure
    // { a: b: { _range: [0, 100] } } ==> a.b._range: [0, 100]
    if (typeof value[k] === 'object' && !Array.isArray(value[k])) {
      const filterName = `${k}.${
        Object.keys(flatten(value[k], { full: true }))[0]
      }`;
      if (pendingReset?.current?.has(filterName)) {
        delete value[k];
        pendingReset.current.delete(filterName);
      }
    }
  });
  return value;
};

// if paging, when anything other than the page changes, reset the page to 1
const resetPage = (nextFormValue, prevFormValue) => {
  if (prevFormValue[formPageKey] && prevFormValue[formPageKey] > 1)
    // eslint-disable-next-line no-param-reassign
    nextFormValue[formPageKey] = 1;
};

// function shared by onSubmit and onChange to coordinate view
// name changes
const normalizeValue = (nextValue, prevValue, views, pendingReset) => {
  if (
    nextValue[formViewNameKey] &&
    nextValue[formViewNameKey] !== prevValue[formViewNameKey]
  ) {
    // view name changed, reset view contents from named view
    return viewToFormValue(
      views.find((v) => v.name === nextValue[formViewNameKey]),
    );
  }
  // something else changed

  // clear empty properties
  const result = clearEmpty(nextValue, pendingReset);

  // if we have a view and something changed, clear the view
  if (result[formViewNameKey]) {
    const view = views.find((v) => v.name === result[formViewNameKey]);
    const viewValue = viewToFormValue(view);
    clearEmpty(viewValue, pendingReset);
    if (Object.keys(viewValue).length !== Object.keys(result).length) {
      delete result[formViewNameKey];
    } else if (
      Object.keys(viewValue).some(
        (k) =>
          // allow mismatch between empty and set strings
          viewValue[k] &&
          result[k] &&
          JSON.stringify(result[k]) !== JSON.stringify(viewValue[k]),
      )
    ) {
      delete result[formViewNameKey];
    }
  }

  return result;
};

// 300ms was chosen empirically as a reasonable default
const DEBOUNCE_TIMEOUT = 300;

export const DataForm = ({
  children,
  footer,
  onDone,
  pad,
  updateOn = 'submit',
  ...rest
}) => {
  const { messages, onView, view, views } = useContext(DataContext);
  const { format } = useContext(MessageContext);
  const [formValue, setFormValue] = useState(viewToFormValue(view));
  // special case for range selectors which always have a value.
  // when value returns to its min/max, remove it from view
  // like other properties
  const pendingReset = useRef(new Set());
  const contextValue = useMemo(() => ({ inDataForm: true, pendingReset }), []);
  const debounce = useDebounce(DEBOUNCE_TIMEOUT);

  const onSubmit = useCallback(
    ({ value }) => {
      const nextValue = normalizeValue(value, formValue, views, pendingReset);
      resetPage(nextValue, formValue);
      setFormValue(nextValue);
      onView(formValueToView(nextValue, views));
      if (onDone) onDone();
    },
    [formValue, onDone, onView, views],
  );

  const onChange = useCallback(
    (value, { touched }) => {
      const nextValue = normalizeValue(value, formValue, views, pendingReset);
      resetPage(nextValue, formValue);
      setFormValue(nextValue);
      if (updateOn === 'change') {
        // debounce search
        if (touched[formSearchKey]) {
          debounce(() => () => onView(formValueToView(nextValue, views)));
        } else {
          onView(formValueToView(nextValue, views));
        }
      }
    },
    [debounce, formValue, onView, updateOn, views],
  );

  useEffect(() => setFormValue(viewToFormValue(view)), [view]);

  let content = children;
  if ((footer !== false && updateOn === 'submit') || pad) {
    content = (
      <Box fill="vertical">
        <Box flex overflow="auto" pad={{ horizontal: pad, top: pad }}>
          <Box flex={false}>{content}</Box>
        </Box>
        {footer !== false && updateOn === 'submit' && (
          <Footer
            flex={false}
            // TO DO theme object
            margin={{ top: 'medium' }}
            pad={{ horizontal: pad, bottom: pad }}
            // TO DO theme object
            gap="small"
          >
            <Button
              label={format({
                id: 'dataForm.submit',
                messages: messages?.dataForm,
              })}
              type="submit"
              primary
            />
          </Footer>
        )}
      </Box>
    );
  }

  return (
    <MaxForm
      {...rest}
      value={formValue}
      onSubmit={updateOn === 'submit' ? onSubmit : undefined}
      onChange={onChange}
    >
      <DataFormContext.Provider value={contextValue}>
        {content}
      </DataFormContext.Provider>
    </MaxForm>
  );
};
