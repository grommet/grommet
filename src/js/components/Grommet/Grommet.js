import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  deviceResponsive,
  getBreakpoint,
  normalizeColor,
  useForwardedRef,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const defaultOptions = {};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    options = defaultOptions,
    messages: messagesProp,
    onAnalytics,
    ...rest
  } = props;
  const { background, dir, themeMode, userAgent } = props;
  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (
      themeProp &&
      themeProp.menu &&
      themeProp.menu.drop &&
      themeProp.menu.drop.align
    ) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';

    if (
      themeMode === 'auto' &&
      typeof window !== 'undefined' &&
      window.matchMedia &&
      window.matchMedia('(prefers-color-scheme: dark)').matches
    ) {
      nextTheme.dark = true;
    }

    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  const messages = useMemo(() => {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    const nextMessages = deepMerge(
      defaultMessages,
      messagesProp?.messages || {},
    );
    return {
      messages: nextMessages,
      format: (opts) => {
        const message = messagesProp?.format && messagesProp.format(opts);
        return typeof message !== 'undefined'
          ? message
          : format(opts, nextMessages);
      },
    };
  }, [messagesProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  const grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  const roots = useRef([]);
  useEffect(() => {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  const rootsContextValue = useMemo(() => ({ roots }), []);

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <RootsContext.Provider value={rootsContextValue}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <OptionsContext.Provider value={options}>
              <MessageContext.Provider value={messages}>
                <AnalyticsProvider onAnalytics={onAnalytics}>
                  <StyledGrommet full={full} {...rest} ref={grommetRef}>
                    {children}
                  </StyledGrommet>
                  {full && <FullGlobalStyle />}
                </AnalyticsProvider>
              </MessageContext.Provider>
            </OptionsContext.Provider>
          </ContainerTargetContext.Provider>
        </RootsContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;

export { Grommet };
