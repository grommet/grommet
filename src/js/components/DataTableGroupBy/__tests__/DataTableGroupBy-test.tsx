import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { hpe } from 'grommet-theme-hpe';

import { Data } from '../../Data';
import { Grommet } from '../../Grommet';
import { DataTableGroupBy } from '..';

const data = [
  { name: 'aa', enabled: true, rating: 2.3, type: { name: 'ZZ', id: 1 } },
  { name: 'bb', enabled: false, rating: 4.3, type: { name: 'YY', id: 2 } },
  { name: 'cc', type: { name: 'ZZ', id: 1 } },
];

describe('DataTableGroupBy', () => {
  test('renders', () => {
    const { container } = render(
      <Grommet theme={hpe}>
        <Data data={data}>
          <DataTableGroupBy
            options={[{ property: 'type.name', label: 'Type' }]}
          />
        </Data>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
