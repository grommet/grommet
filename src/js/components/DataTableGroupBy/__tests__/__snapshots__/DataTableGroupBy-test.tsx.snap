// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableGroupBy renders 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  justify-content: space-between;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 1 1;
  flex-basis: auto;
}

.c10 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-left: 12px;
  margin-right: 12px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c11 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #7D4CDB;
  stroke: #7D4CDB;
}

.c11 g {
  fill: inherit;
  stroke: inherit;
}

.c11 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c11 *[stroke*='#'],
.c11 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c11 *[fill-rule],
.c11 *[FILL-RULE],
.c11 *[fill*='#'],
.c11 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c3 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus >circle,
.c3:focus >ellipse,
.c3:focus >line,
.c3:focus >path,
.c3:focus >polygon,
.c3:focus >polyline,
.c3:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus ::-moz-focus-inner {
  border: 0;
}

.c3:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible) >circle,
.c3:focus:not(:focus-visible) >ellipse,
.c3:focus:not(:focus-visible) >line,
.c3:focus:not(:focus-visible) >path,
.c3:focus:not(:focus-visible) >polygon,
.c3:focus:not(:focus-visible) >polyline,
.c3:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c3:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c8 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
  border: none;
  outline: none;
  border: none;
}

.c8::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c8::-moz-placeholder {
  color: #AAAAAA;
}

.c8:-ms-input-placeholder {
  color: #AAAAAA;
}

.c8 ::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c8::-moz-focus-inner {
  border: none;
  outline: none;
}

.c8:-moz-placeholder,
.c8::-moz-placeholder {
  opacity: 1;
}

.c7 {
  position: relative;
  width: 100%;
}

.c9 {
  cursor: pointer;
}

.c4 {
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c10 {
    margin-left: 6px;
    margin-right: 6px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    id="data"
  >
    <form
      class="c2"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Group by"
        class="c3 c4"
        id="data--groupby"
        type="button"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            >
              <input
                autocomplete="off"
                class="c8 c9"
                id="data--groupby__input"
                inert=""
                name="_groupBy"
                placeholder="Group by"
                readonly=""
                tabindex="-1"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="c10"
            style="min-width: auto;"
          >
            <svg
              aria-label="FormDown"
              class="c11"
              viewBox="0 0 24 24"
            >
              <path
                d="m18 9-6 6-6-6"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </button>
    </form>
  </div>
</div>
`;
